### ISSUE TITLE
Use separate sequence per entity type with Panache

### ISSUE THREAD

**Issue by gunnarmorling** [quarkusio/quarkus#15433]
As [dicussed](https://groups.google.com/g/quarkus-dev/c/-VH3F6iN-Nc) on the mailing list: a separate sequence provides distinct "id spaces" for each entity type, e.g. simplifying tests, also reducing contention during id collection of several busy entity types potentially.

**Comment by gsmet** [comment]
@Sanne I also think that would be a better default (always struggled with that) but is it something supported by Hibernate ORM? I remember having to define sequences for all my classes when I wanted to have this behavior back in the time but maybe I missed something back in the time.

Also, if supported, changing the default will break all the apps out there (and in unexpected ways).

**Comment by gunnarmorling** [comment]
> is it something supported by Hibernate ORM

Isn't this a question of Panache applying the right implicit configuration under the hood? Like it already makes the opinionated assumption that the id is of type long, so it also could apply that sequence config? But I'm super-far away from all this, so it may well be there's some challenge here.

> Also, if supported, changing the default will break all the apps out there (and in unexpected ways).

Yes, that was on my mind, too. Might be good to use sequence-per-type by default but have an option to keep the existing behavior.

**Comment by Sanne** [comment]
Good points about the backwards compatibility; the default Id type being long and its options are provided by the parent class (which is optional to use).

So a simple solution would be to have different parent classes for users to choose from. But also, since you don't have to use the parent class, it's trivial for end users to define their own.

**Comment by gsmet** [comment]
@gunnarmorling the thing is that I'm not sure it's easy to have a working multi-sequences configuration supported by all databases around. That's why I ask if Hibernate ORM makes it possible.

**Comment by Sanne** [comment]
yea it's not an easy one, but considering we have the DB details et al I do agree that Panache should (programmatically?) use a better default - that's Panache's role.

**Comment by loicmathieu** [comment]
We use the `AUTO` ID generation strategy. This strategy usually used a single sequence for the all database.

If we want to use a more sensible default, for example TABLE or SELECT, they may need to be configured, and I'm not even sure we can have the same generator name for multiple entities which will occurs as the generator name would be configured on the base entity.

And even if it's feasible, it would be a backward incompatible change, as id will start back to 0 after the change.

Maybe we need to add a note about the default id strategy to give awardness on this, but I don't think changing it will be an easy (or even doable) task.

**Comment by gsmet** [comment]
> If we want to use a more sensible default, for example TABLE or SELECT

`TABLE` is definitely not a more sensible default. It's far inferior to using sequences when they are available.

**Comment by Sanne** [comment]
Perhaps the right thing to do is propose a change for ORM6 to use a different default for `@GeneratedValue`: in case sequences are available, use a sequence for each object hierarchy (not the same sequence for all entities).

**Comment by gsmet** [comment]
Yeah, that would have my preference (and probably have a config knob for a while to get back to the old behavior).

**Comment by FroMage** [comment]
IMO if the users don't care where the ID is coming from, they really don't. In my case, I couldn't care less if the ID comes from one sequence, more sequences, or if it had to be fetched by a genetically engineered breed of special mice that could come up with unique IDs faster than sequences, provided there was always enough cheese in the database.

I think this is a non-issue. People that have meaningful IDs won't use auto-generated IDs. And people who use auto-generated IDs only happen to use sequences because that's the default for ORM, probably because it's the simplest.

Now, if there are performance reasons that make multiple sequences desirable, we could discuss it, but even then, I feel this is up to ORM to provide the feature.

So, I'm not opposed to this being done in Panache if you insist, but I really don't see the point in the first place.

Am I missing something, @gavinking ?

**Comment by Sanne** [comment]
@FroMage of course users care about where the ID is coming from, and having a separate sequence for each entity hierarchy makes a lot of sense to be used as a better default.

But don't worry, I believe the conclusion is that this isn't a Panache issue; we should propose the new default for Hibernate ORM 6 - as it makes sense beyond Panache, and it's a breaking change.

**Comment by Sanne** [comment]
Just a thought, but I guess the point to which Panache comes into place here is that it's kinda suggesting the notion of "users don't need to worry about it as we'll magically do the right thing" - but then it falls a bit short on what the right thing is, so in that sense yea while we should fix it in ORM, I can see Gunnar's point of raising this as a surprising effect from Panache.

**Comment by FroMage** [comment]
I guess I really don't agree "users care about where the ID is coming from, and having a separate sequence for each entity hierarchy makes a lot of sense to be used as a better default" from my own personal experience.

Now, if everyone agrees with this, then sure, we can support this by default, but that's just going to be because I trust you guys with your opinion. I don't see any fact listed here that can support the assumption that one strategy is better for users of auto-generated IDs, beyond performance points.

For all I care, from a user's PoV, they could be UUIDs if someone told me they were faster to generate and still collision-free.

Even worse, in my case, I would be fine if the DB would give me the generated ID if it supported it (IIRC mysql can do that).

**Comment by Sanne** [comment]
Some clarifications:
 - sure performance is IMO a good argument, but N.B. Gunnar reported this because he was seeing unexpected gaps in the sequence; so it's also surprising behaviour.
 - sure MySQL can do auto-increment, and we do use that instead of Sequences if you're using MySQL..
 - UUIDs are actually a terrible choice from a performance point of view - with some exceptions of course, as they might make sense on certain cases - but you're most likely going to deplete the entrophy pool on the server if you abuse UUIDs, at which point the whole server gets extremely slow; not to mention their size, which is expensive to handle in all stages: within the JVM, to serialize, on the wire, and storage efficiency.

**Comment by gavinking** [comment]
But I thought this was what `hibernate.jpa.compliance.global_id_generators` does? @sebersole

**Comment by gavinking** [comment]
> having a separate sequence for each entity hierarchy makes a lot of sense to be used as a better default.

Pretty sure everyone except for the JPA spec agrees on this. I'm not even sure how it was that the JPA spec came to define something different.

**Comment by FroMage** [comment]
> sure MySQL can do auto-increment, and we do use that instead of Sequences if you're using MySQL..

So, if `AUTO` means on some DBs it's a sequence and on others it's what the DB provides, that sorta proves my point that users don't care where the damn ID is coming from, no?

> he was seeing unexpected gaps in the sequence

IMO, this appears to be an OCD issue ;) Nobody actually cares about sequence gaps, unless they lead to depletion, but that's not the case here.

**Comment by gavinking** [comment]
> sure MySQL can do auto-increment, and we do use that instead of Sequences if you're using MySQL.

Which is another reason the JPA spec is broken on this point: on some platforms `AUTO` means unique-per-table, and on others it means unique-per-PU.

That's actually quite terrible.

**Comment by FroMage** [comment]
But OK, if we're going to change this, whether in Panache or in ORM, how are we going to be able to not break existing applications if we change the default?

Introducing a new supertype sounds like introducing confusion for opting into a new default that again, even if it's better design, I really doubt any user of `AUTO` cares about.

**Comment by gavinking** [comment]
> So, if `AUTO` means on some DBs it's a sequence and on others it's what the DB provides, that sorta proves my point that users don't care where the damn ID is coming from, no?

More likely it means that they have to avoid using `AUTO` for real programs, and explicitly specify stuff.

**Comment by FroMage** [comment]
Again, my point is that you never have that issue for natural IDs, or for people who care about their IDs and are not using `AUTO`.

**Comment by gavinking** [comment]
https://hibernate.atlassian.net/browse/HHH-12329?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel

> > That's actually quite terrible.
> 
> I don't know if it is. To be the semantics of `AUTO` is "I don't give a shit, just gimme an unique ID", and that behaviour fits the goal, no?

YOLO is, in my experience, a pretty terrible principle to bring to the practice of database design.

**Comment by gavinking** [comment]
> But OK, if we're going to change this, whether in Panache or in ORM, how are we going to be able to not break existing applications if we change the default?

But again, and maybe I'm just wrong, but *isn't this what `hibernate.jpa.compliance.global_id_generators` does*??? This isn't something new at all.

**Comment by gavinking** [comment]
> Excuse me, that's wrong.
> 
> I mean `hibernate.model.generator_name_as_sequence_name`.

And that doesn't help if you don't specify a generator name explicitly.

**Comment by gunnarmorling** [comment]
Fixing this in either ORM itself or Panache would be fine from my PoV. I opened this request here as I consider Panache to provide a more prescriptive, opinionated layer with sensible defaults. And sequence-per-hierarchy surely is such better default, for the reasons pointed out by @gavinking and @Sanne. Agreed that assuming no gaps isn't something you'd do in prod, but for tests it's fine to rely on expected ids as your test inserts entities.

**Comment by gavinking** [comment]
Yeah, forget it, I'm wrong. There doesn't seem to be a way to make `AUTO` not use this nasty `hibernate_sequence` thing.

Personally that seems pretty bad to me.

**Comment by gavinking** [comment]
> Fixing this in either ORM itself

We should at least fix it in H6. Who wants to start a [Discussion](https://github.com/hibernate/hibernate-orm/discussions)?

**Comment by Sanne** [comment]
> > That's actually quite terrible.
> 
> I don't know if it is. To be the semantics of `AUTO` is "I don't give a shit, just gimme an unique ID", and that behaviour fits the goal, no?

The difference with Panache is that Hibernate ORM will ask users to explicitly to subscribe to such a choice: sure there's some people that couldn't care less, and they have the _option_ to use `AUTO`. Very useful in quick throw-away POCs, but it's also useful that, as a user :
 a) I can easily evolve it later
 b) I made the choice myself - it's not hidden and surprising

If Panache's goal is to be magical and remove the needs for such concerns, then it needs to do the right thing for all users.

**Comment by sebersole** [comment]
I just skimmed because there is too much to read and unwrap here.  But I saw a few themes...

AUTO means, as the spec states, that the provider is free to choose what to do.  It also implies that the user in fact does not care in terms of the actual strategy.  I am very happy with what Hibernate does right now in terms of the strategies it selects and will not be changing that.

The name `hibernate_sequence` sucks - absolutely.  Its not like that just started sucking.  Its sucked for that last 20 years.  I fully appreciate and am largely jealous that y'all get to not worry about backwards compatibility, but unfortunately ORM does.  Changing that behavior *has to* have a proper way to adjust existing schemas.  

Changing the strategy for determining the implicit name of the database structure used to back the generator is similar to the generator `prefer_sequence_per_entity` config value.  That setting says to use the entity name as the sequence name as opposed to `hibernate_sequence`.  What might be nicer is a pluggable strategy there.  The difficulty there is passing the right information; because we generate ids for entities as well as id-bags its not like we can just pass the PersistentClass graph for the entity.

**Comment by StephenOTT** [comment]
Any movement on this?
